core/src/main/java/com/linecorp/armeria/client/retry/RetryingClient.java

```diff
@@ -65,4 +84,66 @@ protected Backoff newBackoff() {
     protected RetryStrategy<I, O> retryStrategy() {
         return retryStrategy;
     }
+
+    /**
+     * Returns the number of requests sent and increase its count by one.
+     * This method should be invoked when the first request is made.
+     */
+    protected final int increaseNumRequestsSent(ClientRequestContext ctx) {
+        if (ctx.responseTimeoutMillis() <= 0 || !responseTimeoutEnabled) {
+            responseTimeoutEnabled = false;
+        } else {
+            if (originalResponseTimeout == -1) {
+                originalResponseTimeout = ctx.responseTimeoutMillis();
+            }
+            lastTimeoutAndTimestamp.put(
+                    numRequestsSent, new LastTimeoutAndTimestamp(
+                            originalResponseTimeout, TimeUnit.NANOSECONDS.toMillis(System.nanoTime())));
+        }
+        return numRequestsSent++;
+    }
+
+    /**
+     * Resets the {@link ClientRequestContext#responseTimeoutMillis()}.
+     */
+    protected final void resetResponseTimeout(ClientRequestContext ctx, final int reqId) {
+        if (!responseTimeoutEnabled) {
+            return;
+        }
+
+        final long responseTimeoutMillis = responseTimeoutMillis(reqId);
+        if (responseTimeoutMillis <= 0) { // 0 is not from the first, but subtracted to that, so it's timeout.
+            throw ResponseTimeoutException.get();
+        }
+
+        final LastTimeoutAndTimestamp timeoutAndStamp = this.lastTimeoutAndTimestamp.get(reqId);
+        timeoutAndStamp.lastTimeoutMillis = responseTimeoutMillis;
+        timeoutAndStamp.lastTimestampMills = TimeUnit.NANOSECONDS.toMillis(System.nanoTime());
+        ctx.setResponseTimeoutMillis(responseTimeoutMillis);
+    }
+
+    /**
+     * Returns the remaining response timeout milliseconds or -1 if response timeout is disabled.
+     */
+    protected final long responseTimeoutMillis(final int reqId) {
+        if (!responseTimeoutEnabled) {
+            return -1;
+        }
+
+        long responseTimeoutMillis =
+                lastTimeoutAndTimestamp.get(reqId).lastTimeoutMillis -
+                (TimeUnit.NANOSECONDS.toMillis(System.nanoTime()) -
+                 lastTimeoutAndTimestamp.get(reqId).lastTimestampMills);
```

Comment:
This seems more complex than it can be. How about using a dead line instead?

```java
// When creating a LastTimeoutAndTimestamp:
lastTimeoutAndTimestamp.deadlineNanos = lastTimeoutAndTimestamp.startTimeNanos + responseTimeoutMillis * 1000_000;
// To calculate the response timeout:
long responseTimeoutMillis = (lastTimeoutAndTimestamp.deadlineNanos - System.nanoTime()) / 1000_000;
if (responseTimeoutMillis <= 0) { ... timed out ... }
```

That being said, lastTimeoutAndTimestamp could be replaced with a single long integer holder?